indico/modules/rb_new/client/js/modules/admin/EditableListItem.jsx

```diff
@@ -0,0 +1,134 @@
+/* This file is part of Indico.
+ * Copyright (C) 2002 - 2018 European Organization for Nuclear Research (CERN).
+ *
+ * Indico is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * Indico is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Indico; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+import _ from 'lodash';
+import React from 'react';
+import PropTypes from 'prop-types';
+import {Form as FinalForm} from 'react-final-form';
+import {Button, Confirm, Form, List} from 'semantic-ui-react';
+import {Translate} from 'indico/react/i18n';
+
+import './EditableList.module.scss';
+
+
+/**
+ * A list item component that supports inline editing and deletion.
+ */
+export default class EditableListItem extends React.PureComponent {
+    static propTypes = {
+        item: PropTypes.object.isRequired,
+        renderDisplay: PropTypes.func.isRequired,
+        renderEditForm: PropTypes.func.isRequired,
+        confirmDeleteMessage: PropTypes.any.isRequired,
+        initialEditValues: PropTypes.func,
+        onDelete: PropTypes.func.isRequired,
+        onUpdate: PropTypes.func.isRequired,
+    };
+
+    static defaultProps = {
+        initialEditValues: item => item,
+    };
+
+    state = {
+        // delete
+        confirming: false,
```

Comment:
Wouldn't it be better to have sth like `state = {delete: {state: ...}, edit: {state: ...}}`?

Reply:
react doesn't do deep merges for state changes so i'd always have to specify the whole object